socs/agents/generator/agent.py

```diff
@@ -0,0 +1,357 @@
+            def evaluator(registers): return interp_unsigned_double_reg(registers[offset], registers[offset + 1])
+        elif rconfig['read_as'] == '32S':
+            def evaluator(registers): return interp_signed_double_reg(registers[offset], registers[offset + 1])
+        else:
+            def evaluator(registers): return self.error_val
+
+        def process(registers):
+            val = evaluator(registers)
+            if 'scale' in rconfig:
+                val = val * rconfig['scale']
+
+            if self.error_out_of_range:
+                try:
+                    if val < rconfig['min_va']:
+                        val = self.error_val
+                except KeyError:
+                    pass
+
+                try:
+                    if val > rconfig['max_val']:
+                        val = self.error_val
+                except KeyError:
+                    pass
+
+            return {rconfig['name'].replace(' ', '_'): {'value': val, 'units': rconfig['units']}}
+
+        return process
+
+    def read(self, client):
+        regdata = client.read_holding_registers(self.read_start, self.read_len)
+        regs = regdata
+        return_data = {}
+        try:
+            for i in range(self.read_len):
+                return_data.update(self.functions[i](regs))
+        except Exception as e:
+            print(f'Error in processing data: {e}')
+
+        return return_data
+
+
+class GeneratorAgent:
+
+    def __init__(self, agent, host='localhost', port=5021, read_config=None):
+
+        self.host = host
+        self.port = port
+        self.read_strings = []
+
+        self.agent: ocs_agent.OCSAgent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+
+        self.initialized = False
+        self.take_data = False
+
+        self.client = None
+
+        if read_config is not None:
+            with open(read_config) as f:
+                data = yaml.load(f, Loader=yaml.FullLoader)
+            self.build_config([data])
+            self.log.info("Config file loaded successfully.")
+        else:
+            self.log.info("No config file provided.")
+
+        agg_params = {
+            'frame_length': 10 * 60  # [sec]
+        }
+        self.agent.register_feed('generator',
+                                 record=True,
+                                 agg_params=agg_params,
+                                 buffer_time=0)
+
+    def build_config(self, config):
+        for string in config:
+            self.read_strings.append(ReadString(string))
+
+    def read_cycle(self, client):
+        for i, val in enumerate(self.read_strings):
+            data = self.read_regs(client, val)
+            return data
+
+    def read_regs(self, client, register_string_obj):
+        try:
+            data = register_string_obj.read(client)
+        except Exception as e:
+            print('error in read', e)
+            return
+
+        return data
```

Comment:
These three methods (and their associated attributes) seem like they would fit nicely in a separate object that is used to communicate with the generator. Perhaps this object could also manage the `ModbusClient`, since this is where that gets used to. I think separating this out would clean up some of the logic a bit from the perspective of the Agent.